pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+
+			// get the bringe-X dns request/replies
+			dnssys[bnNum].Lock()
+			defer dnssys[bnNum].Unlock()
+			for _, dnsdata := range dnssys[bnNum].Snoop {
+				// unique by domain name, latest reply overwrite previous ones
+				if dnsdata.isIPv4 {
+					dnsrec[0][dnsdata.DomainName] = dnsdata
+				} else {
+					dnsrec[1][dnsdata.DomainName] = dnsdata
+				}
+			}
+			for idx := range dnsrec {
+				for _, dnsRec := range dnsrec[idx] {
+					// temp print out all unique dns replies for the bridge
+					log.Infof("!!FlowStats: DNS time %v, domain %s, appIP %v, count %d, Answers %v",
+						dnsRec.TimeStamp, dnsRec.DomainName, dnsRec.AppIP, dnsRec.ANCount, dnsRec.Answers)
+
+					dnsrec := types.DNSReq{
+						HostName:    dnsRec.DomainName,
+						Addrs:       dnsRec.Answers,
+						RequestTime: dnsRec.TimeStamp.UnixNano(),
+					}
+					dnsPacked = true
+					flowdata.DNSReqs = append(flowdata.DNSReqs, dnsrec)
+				}
+			}
+			// flow record done for the bridge/app
+			//flowdata.CastFlowStatus
+			flowKey := scope.UUID.String() + scope.NetUUID.String()
+			pub.Publish(flowKey, &flowdata)
+		}
+	}
+
+	// remove the dns data already uploaded
+	if dnsPacked {
+		for bnx := range instData.bnNet {
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+			dnssys[bnNum].Snoop = nil
+			log.Infof("!!FlowStats: clear dns record for bn%d", bnNum)
+		}
+	}
+}
+
+// conntrack flow of two uni-directional stats into one
+// bireditional flow stats
+func flowMergeProcess(entry *netlink.ConntrackFlow, instData networkAttrs) flowStats {
+	var ipFlow flowStats
+	var forwSrcApp, forwDstApp, backSrcApp, backDstApp bool
+	var AppNum int
+
+	// for the current phase of implementation, we ignore the flowstats which has not timed out yet
+	ipflowTimeOut := int32(entry.TimeOut)
+	if ipflowTimeOut > timeoutSec {
+		return ipFlow
+	}
+
+	ipFlow.aclNum = entry.Mark & markMask
+	ipFlow.appNum = uint8(entry.Mark >> appShiftBits)
+	AppNum = int(ipFlow.appNum)
+
+	//AppNum := 1 // XXX hack, = ipFlow.appNum later
+	//ipFlow.appNum = uint8(AppNum)
+	ipFlow.TimeStart = entry.TimeStart
+	ipFlow.TimeStop = entry.TimeStop
+	ipFlow.TimeOut = entry.TimeOut
+	ipFlow.Proto = entry.Forward.Protocol
+	ipFlow.IsTimeOut = true
+
+	// Assume the App has an assigned IP address(es) first
+	if len(instData.appIPAddrs[AppNum]) > 0 {
+		ipFlow.dbg1 = 1
+		forwSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.SrcIP)
+		if forwSrcApp == false {
+			ipFlow.dbg1 = 2
+			forwDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.DstIP)
+			if forwDstApp == false {
+				ipFlow.dbg1 = 3
+				backSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.SrcIP)
+				if backSrcApp == false {
+					ipFlow.dbg1 = 4
+					backDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.DstIP)
+				} else {
+					ipFlow.dbg1 = 5 // XXX
+				}
+			}
+		}
+	}
+
```

Comment:
The logic below is not very obvious when read for the first time. Might make sense to explain the rationale behind different if else cases in comments.

Reply:
will add some more comments on this part